#31363 format template configure in settings

Blargian opened the pull request:
Hi Clickhouse team, 
I'm opening a PR for this feature. Look forward to your feedback and fixing any issues. :) 

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The user can now specify the template string directly in the query using `format_schema_rows_template` as an alternative to `format_template_row`. Closes https://github.com/ClickHouse/ClickHouse/issues/31363.

### Documentation entry for user-facing changes

- [X] Documentation is written - English and Russian. (То что написал на русском скорее всего есть ошибки, я не носитель русского... нужно помощь исправить пж)

- docs/en/interfaces/formats.md
- docs/ru/interfaces/formats.md
- docs/en/operations/settings/settings-formats.md

Example usage not yet added, I can add that if needed.

Blargian committed:
added format_schema_rows_template setting

Blargian committed:
#31363 - modified TemplateBlockOutputFormat to work with added format_schema_rows_template setting

Blargian committed:
#31363 - remove schema delimiter setting and add test 00937_format_schema_rows_template.sh and reference

Blargian committed:
#31363 - update documentation for En and Ru

Blargian committed:
empty commit to restart CI checks

Felixoid commented:
Dear @Blargian, the `#` in the branch name breaks the CI because we don't URL-encode it. The fix won't be instant, so please, recreate the PR from another branch.

Avogar commented:
The problem with CI is the branch name, we have a bug in python script with URL encoding (symbol `#` in the branch name wasn't encoded properly). We will fix it, but it will take time. You can recreate a PR with new branch name or wait until it's fixed

robot-ch-test-poll2 commented:
<!-- automatic status comment for PR #59088 from Blargian/ClickHouse:#31363_format_template_configure_in_settings -->
*This is an automated comment for commit 5f8565307e40a24247bbdd34cb9c9097f762af31 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59088/5f8565307e40a24247bbdd34cb9c9097f762af31/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59088/5f8565307e40a24247bbdd34cb9c9097f762af31/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

Felixoid commented:
I have restored the previous commit to not affect the autoapprove lambda. It should block changes into `tests/ci` directory

Blargian committed:
fix typo in formats.md

Blargian committed:
Update 00937_format_schema_rows_template.sh

Blargian commented:
@Felixoid thanks for the heads up and also for the fix!

Felixoid committed:
Merge branch 'master' into #31363_format_template_configure_in_settings

Felixoid commented:
> Do you maybe know why?

Do you know a joke?

> -- We have good and bad news for you
> -- I'd like to start with a good part, doctor
> -- We'll call decease after you

https://github.com/ClickHouse/ClickHouse/pull/59184

Blargian commented:
> Do you know a joke?
> 
> > -- We have good and bad news for you
> > -- I'd like to start with a good part, doctor
> > -- We'll call decease after you

Рад служить! )

Blargian committed:
Update 00937_format_schema_rows_template.sh

fix failing shellcheck

Blargian committed:
fix failing 00937_template_output_format

Blargian committed:
fix failing test 00937_format_schema_rows_template.sh

Blargian committed:
Merge branch 'ClickHouse:master' into #31363_format_template_configure_in_settings

Blargian committed:
CI trigger

Avogar commented on src/Core/Settings.h:
```diff
@@ -1080,6 +1080,7 @@ class IColumn;
     M(String, format_schema, "", "Schema identifier (used by schema-based formats)", 0) \
     M(String, format_template_resultset, "", "Path to file which contains format string for result set (for Template format)", 0) \
```
Let's also add similar setting for resultset template, will be useful to have both

Avogar replied:
Also, please, add new settings to `SettingsChangesHistory.h`

Avogar commented on src/Processors/Formats/Impl/TemplateBlockOutputFormat.cpp:
```diff
@@ -213,14 +214,27 @@ void registerOutputFormatTemplate(FormatFactory & factory)
+        if (settings.template_settings.row_format.empty())
+        {
+            row_format = ParsedTemplateFormatString();
+            row_format.parse(settings.template_settings.row_format_schema,idx_by_name);
```
```suggestion
            row_format.parse(settings.template_settings.row_format_schema, idx_by_name);
```

Avogar commented on src/Core/Settings.h:
```diff
@@ -1080,6 +1080,7 @@ class IColumn;
     M(String, format_schema, "", "Schema identifier (used by schema-based formats)", 0) \
     M(String, format_template_resultset, "", "Path to file which contains format string for result set (for Template format)", 0) \
     M(String, format_template_row, "", "Path to file which contains format string for rows (for Template format)", 0) \
+    M(String, format_schema_rows_template, "", "Format string for rows (for Template format)", 0) \
```
Let's rename it to `format_template_row_schema` or `format_template_row_format`. We should keep `format_template_` prefix as it indicates that this setting is related only to format Template

Blargian committed:
rename of settings, add setting for resultset, extend test, fix documentation and add to SettingsChanges log

Blargian committed:
Merge branch 'master' into #31363_format_template_configure_in_settings

Avogar commented on src/Core/SettingsChangesHistory.h:
```diff
@@ -102,6 +102,8 @@ static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> sett
               {"function_visible_width_behavior", 0, 1, "We changed the default behavior of `visibleWidth` to be more precise"},
               {"max_estimated_execution_time", 0, 0, "Separate max_execution_time and max_estimated_execution_time"},
               {"iceberg_engine_ignore_schema_evolution", false, false, "Allow to ignore schema evolution in Iceberg table engine"},
+              {"format_template_row_format", "none", "", "Template row format string can be set directly in query"},
+              {"format_template_resultset_format", "none", "", "Template result set format string can be set in query"},
```
```suggestion
              {"format_template_row_format", "", "", "Template row format string can be set directly in query"},
              {"format_template_resultset_format", "", "", "Template result set format string can be set in query"},
```
No need to use "none", just keep empty

Blargian committed:
Update src/Core/SettingsChangesHistory.h

set previous value as empty string for added setting rather than "none" in SettingsChangesHistory.h

Co-authored-by: Kruglov Pavel <48961922+Avogar@users.noreply.github.com>

Blargian commented:
@Avogar the last few tests are all failing on issues related to not being able to find, or remove a `.tmp` version of my shell test. Where does the `.tmp` come from? Is there a specific permission level that test files should be given?

Avogar commented:
The problem is that you used filenames `00937_template_output_format_row.tmp//00937_template_output_format_resultset.tmp`, the same as in test `00937_template_output_format.sh`, but we run tests in parallel, so one test can create files, another remove them. Also in `Stateless tests flaky check` we run the same test in parallel with itself, so the same problem. Let's use unique filenames using env variable `$CLICKHOUSE_TEST_UNIQUE_NAME`

Blargian committed:
update test to use CLICKHOUSE_TEST_UNIQUE_NAME so parallel tests don't fail

Blargian committed:
2 tests fail, doesnt seem change related - try again

Blargian committed:
Merge branch 'master' into #31363_format_template_configure_in_settings

Blargian commented:
@Avogar is there some bigger issue with CI at the moment? Fast tests fails since I updated my branch from master this morning. It was okay yesterday.

Blargian committed:
Merge remote-tracking branch 'upstream/master' into #31363_format_template_configure_in_settings

Avogar commented:
Test 02720_row_policy_column_with_dots was broken in master by accident, already fixed, so should be ok now

Blargian committed:
trigger CI again

Blargian committed:
fix failing upgrade check due to identical previous_value and new_value

Blargian committed:
change back SettingsChangesHistory to empty strings - not the reason for CI failing on upgrade check

Blargian committed:
Merge branch 'master' into #31363_format_template_configure_in_settings

Blargian committed:
Update SettingsChangesHistory.h

Fix failing CI upgrade check due to setting changes being listed in v24.1 when the latest version has advanced to 24.2 after merging in master branch

Blargian committed:
Merge branch 'ClickHouse:master' into #31363_format_template_configure_in_settings

Blargian commented:
**Remaining tests failing**:

- test_storage_kafka/test.py::test_system_kafka_consumers_rebalance_mv (#56043)
- 02981_vertical_merges_memory_usage (#59648)

Avogar commented:
Sorry for waiting, I was on vacation. I will resolve conflicts and merge this PR when tests are finished

Avogar committed:
Merge branch 'master' into #31363_format_template_configure_in_settings

Blargian commented:
> Sorry for waiting, I was on vacation. I will resolve conflicts and merge this PR when tests are finished

No worries, thanks! Hope it was a good vacation :+1:
